### PR TITLE
[NavigationDrawer] Added ability to account for bottom safe area in presentation of initial drawer height

### LINKED ISSUE #9322: [NavigationDrawer] Internal issue: b/146566784
This was filed as an internal issue. If you are a Googler, please visit [b/146566784](http://b/146566784) for more details.

---
#### Internal data
- Associated internal bug: [b/146566784](http://b/146566784)

### PR DESCRIPTION
**Pull request by jakerockland**
Added ability to account for bottom safe area in presentation of initial drawer height, parameterized by default-off flag.

Fixes #9322.

I tried adding unit tests on this, per our conversation @yarneo, but ran into some issues with writing tests around this as: 
1) unit testing for this is difficult as one cannot manually set the `safeAreaInsets` 
2) snapshot testing this requires that the test suite is done on an iPhone X or similar device with bottom insets, which is not the current test device for snapshots as far as I can tell 

I'm happy to add tests but need a bit of guidance regarding how to best test this given the above constraints. I didn't see a test testing the related `shouldIncludeSafeAreaInContentHeight` behavior, which I was looking to use as a test reference, but perhaps could do something similar to: https://github.com/material-components/material-components-ios/blob/b8090cb6382910cc4190c7030739e145f3856929/components/FlexibleHeader/tests/unit/supplemental/FlexibleHeaderTopSafeAreaTestsFakeViewController.m#L90-L94.

Wanted to get your thoughts on this before diving further @yarneo.

### CODE REVIEW

**Comment by jverkoey** [comment]
b/146566784 maps to https://github.com/material-components/material-components-ios/issues/7169

**Comment by jverkoey** [comment]
I've fixed that; it looks like the internal issue may have been copied from another issue resulting in two issues pointing to the same GitHub issue.

**Comment by jakerockland** [comment]
@jverkoey sorry about that, I think I did use a different issue as a template and must not have cleared out the Github field 😓

**Comment by jverkoey** [comment]
No worries! That field really just shouldn't be getting copied when an issue is used as a template.

**Comment by yarneo** [comment]
Thanks @jakerockland ! Will take a look today

**Review by yarneo — COMMENTED** [review]
Thank you for contributing!

I looked at the PR and it looks great overall.

Regarding testing the bottom safe area, i wonder if this would help: https://github.com/material-components/material-components-ios/blob/develop/components/FlexibleHeader/tests/unit/supplemental/FlexibleHeaderTopSafeAreaTestsFakeViewController.m#L90

Basically overriding the safe area insets of the view and thus imitating an iphoneX

**Comment by jakerockland** [comment]
@yarneo Happy to be contributing here! 😄 

So I tried using https://github.com/material-components/material-components-ios/blob/b8090cb6382910cc4190c7030739e145f3856929/components/FlexibleHeader/tests/unit/supplemental/FlexibleHeaderTopSafeAreaTestsFakeViewController.m as a reference, adding a test in `MDCNavigationDrawerPresentationControllerTests.m` by setting the `MDCBottomDrawerContainerViewController`'s view directly but am not sure if this approach is workable without faking out the `MDCBottomDrawerContainerViewController` itself, which in some way defeats the intent of the test.

This is the approach I tried variations of, which wasn't working for me, though perhaps I was overlooking something silly 😓:

```obj-c
@interface FakeBottomSafeAreaView : UIView
@end

@implementation FakeBottomSafeAreaView

- (UIEdgeInsets)safeAreaInsets {
  UIEdgeInsets safeAreaInsets = [super safeAreaInsets];
  safeAreaInsets.bottom = 34.f;
  return safeAreaInsets;
}

@end

@implementation MDCNavigationDrawerPresentationControllerTests

// ... stuff removed for snippet

- (void)testInitialDrawerHeightBottomSafeArea {
  // Given
  UIViewController *fakeContentViewController = [[UIViewController alloc] init];
  MDCBottomDrawerViewController *fakeBottomDrawer = [[MDCBottomDrawerViewController alloc] init];
  fakeBottomDrawer.contentViewController = fakeContentViewController;
  UIViewController *fakePresentingViewController = [[UIViewController alloc] init];
  MDCBottomDrawerPresentationController *presentationController =
      [[MDCBottomDrawerPresentationController alloc]
          initWithPresentedViewController:fakeBottomDrawer
                 presentingViewController:fakePresentingViewController];
  presentationController.bottomDrawerContainerViewController.view = [[FakeBottomSafeAreaView alloc] init];

  // When
  [presentationController presentationTransitionWillBegin];
  CGFloat height = presentationController.bottomDrawerContainerViewController.maximumInitialDrawerHeight;
  presentationController.shouldIncludeSafeAreaInInitialDrawerHeight = YES;
  CGFloat safeAreaHeight = presentationController.bottomDrawerContainerViewController.maximumInitialDrawerHeight;

  // Then
  XCTAssertTrue(safeAreaHeight > height);
}

// ... stuff removed for snippet

@end
```

A couple tests which are easier to add are just testing the property flag propagation, which may offer sufficient testing, given how simple the logic is for actually adding the bottom inset. 

I've added tests for those to this PR—for both this new flag and the previously existing similar one. 

Thoughts on this?

File: components/NavigationDrawer/tests/unit/MDCNavigationDrawerPresentationControllerTests.m [hunk 1]
```diff
@@ -115,4 +115,45 @@ - (void)testSettingShouldAlwaysExpandHeader {
   XCTAssertTrue(
       presentationController.bottomDrawerContainerViewController.shouldAlwaysExpandHeader);
 }
+
+- (void)testShouldIncludeSafeAreaInContentHeight {
```
Comment (yarneo) [inline components/NavigationDrawer/tests/unit/MDCNavigationDrawerPresentationControllerTests.m:119]: this test looks good, the bottom one looks like a duplicate? Would we be able to test the propagation from the MDCBottomDrawerViewController to the MDCBottomDrawerPresentationController?

**Comment by yarneo** [comment]
We chatted offline, and the propagation route seems reasonable based on the difficulties of imitating the safe area in the unit test for the drawer container view controller

File: components/NavigationDrawer/tests/unit/MDCNavigationDrawerPresentationControllerTests.m (hunk 1 above)
Comment (jakerockland) [reply components/NavigationDrawer/tests/unit/MDCNavigationDrawerPresentationControllerTests.m:119]: So this test is actually just a bit of a bonus test (could be added in a separate PR but seemed relevant here or I can remove as well) which tests the existing `shouldIncludeSafeAreaInContentHeight` flag for content height taking safe areas into account. 


The test below does the actual testing around `shouldIncludeSafeAreaInInitialDrawerHeight`, which is the newly added but similarly named flag.

Testing though that the flag propagates from MDCBottomDrawerViewController to the MDCBottomDrawerPresentationController sounds good to me.

I'm signing off for today but will revisit this on Monday and I can modify both of these tests to set the flag on the MDCBottomDrawerViewController and verify that it propagates to both the MDCBottomDrawerPresentationController and the container VC?

Comment (yarneo) [reply components/NavigationDrawer/tests/unit/MDCNavigationDrawerPresentationControllerTests.m:119]: ohh i see now! Perfect! Yes sounds good to me :)

Comment (jakerockland) [reply components/NavigationDrawer/tests/unit/MDCNavigationDrawerPresentationControllerTests.m:119]: Awesome, thanks @yarneo! I'll make those tweaks to the propagation test then as well as cleaning up the clang-format warning on Monday. Have a great weekend 😄

**Comment by jakerockland** [comment]
Hey @yarneo, I added some additional tests for testing the propagation from the MDCBottomDrawerViewController to the MDCBottomDrawerPresentationController. I added these both for the existing `shouldIncludeSafeAreaInContentHeight` flag as well as the newly added `shouldIncludeSafeAreaInInitialDrawerHeight` one.

**Comment by yarneo** [comment]
Awesome! Looks great! Re-running CI to make sure everything is good to go

**Review by yarneo — APPROVED** [review]

**Comment by yarneo** [comment]
Thanks a bunch!! Merged the PR

**Pull request merged**